### PR TITLE
Implement conditional inserts

### PR DESCRIPTION
**Pull request by olim7t**

### CODE REVIEW

File: testing/src/test/java/io/stargate/it/http/graphql/graphfirst/InsertTest.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            + "type InsertUserResponse @cql_payload {\n"
+            + "  applied: Boolean!\n"
+            + "  user: User!\n"
+            + "}\n"
+            + "type Mutation {\n"
+            + "  insertUser(user: UserInput!): User\n"
+            + "  persistUser(user: UserInput!): User @cql_insert\n"
+            + "  insertUser2(user: UserInput!): InsertUserResponse\n"
+            + "  insertUserIfNotExists(user: UserInput!): InsertUserResponse\n"
+            + "  insertUser3(user: UserInput!): InsertUserResponse @cql_insert(ifNotExists: true)\n"
```
Comment (olim7t) [inline testing/src/test/java/io/stargate/it/http/graphql/graphfirst/InsertTest.java:62]: This sums up the new features added by this PR.

File: testing/src/test/java/io/stargate/it/http/graphql/graphfirst/InsertTest.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            + "  name: String\n"
+            + "  username: String\n"
+            + "}\n"
+            + "type Query { user(id: ID!): User }\n"
+            + "type InsertUserResponse @cql_payload {\n"
+            + "  applied: Boolean!\n"
+            + "  user: User!\n"
+            + "}\n"
+            + "type Mutation {\n"
+            + "  insertUser(user: UserInput!): User\n"
```
Comment (tomekl007) [inline testing/src/test/java/io/stargate/it/http/graphql/graphfirst/InsertTest.java:58]: what is this `UserInput`? I don't see the creation of this type anywhere

File: graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/UpdateMappingModelBuilder.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  @Override
+  MutationMappingModel build() throws SkipException {
+
+    // TODO more options for signature
+    // Currently requiring exactly one argument that must be an entity input with all PK fields set.
+    // We could also take the PK fields directly (need a way to specify the entity), partial PKs for
+    // multi-row deletions, additional IF conditions, etc.
+
+    Type<?> returnType = TypeHelper.unwrapNonNull(mutation.getType());
+    if (!(returnType instanceof TypeName) || !"Boolean".equals(((TypeName) returnType).getName())) {
```
Comment (tomekl007) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/UpdateMappingModelBuilder.java:53]: nit: `"Boolean"` to a constant.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/MutationMappingModelBuilder.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      InputValueDefinition input,
+      Map<String, EntityMappingModel> entities,
+      ProcessingContext context,
+      String mutationName,
+      String mutationKind)
+      throws SkipException {
+
+    Type<?> type = TypeHelper.unwrapNonNull(input.getType());
+
+    if (type instanceof ListType) {
```
Comment (tomekl007) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/MutationMappingModelBuilder.java:43]: why you are checking only `ListType`? As far as I can see, there could be other types (i.e.: `MapType`).
Maybe we should check `if(!type instnanceof TypeName)`?

File: graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/MutationMappingModelFactory.java [hunk 5]
```diff
@@ -67,17 +70,17 @@ private static Kind detectType(FieldDefinition mutation, ProcessingContext conte
    * explicit directive, or otherwise from a set of predefined name prefixes.
    */
   private enum Kind {
-    INSERT(InsertMappingModel::build, "cql_insert", "insert", "create"),
-    UPDATE(UpdateMappingModel::build, "cql_update", "update"),
-    DELETE(DeleteMappingModel::build, "cql_delete", "delete", "remove"),
+    INSERT(InsertMappingModelBuilder::new, "cql_insert", "insert", "create"),
+    UPDATE(UpdateMappingModelBuilder::new, "cql_update", "update"),
+    DELETE(DeleteMappingModelBuilder::new, "cql_delete", "delete", "remove"),
     ;
```
Comment (tomekl007) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/MutationMappingModelFactory.java:76]: why no `SELECT` here?

File: graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/ResponseMappingModel.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /** Whether the field is a list of entities, or a single entity. */
+    public boolean isList() {
+      return isList;
+    }
+  }
+
+  public enum TechnicalField {
+    APPLIED("applied", Scalars.GraphQLBoolean.getName()),
```
Comment (tomekl007) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/ResponseMappingModel.java:86]: just from curiosity, what other `TechnicalField`s are we anticipating?

File: testing/src/test/java/io/stargate/it/http/graphql/graphfirst/InsertTest.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    NAMESPACE = keyspaceId.asInternal();
+    CLIENT.deploySchema(
+        NAMESPACE,
+        "type User @cql_input {\n"
+            + "  id: ID!\n"
+            + "  name: String\n"
+            + "  username: String\n"
+            + "}\n"
+            + "type Query { user(id: ID!): User }\n"
+            + "type InsertUserResponse @cql_payload {\n"
```
Comment (tomekl007) [inline testing/src/test/java/io/stargate/it/http/graphql/graphfirst/InsertTest.java:53]: shouldn't we consider auto-generating this class similar to `*Input` classes when using (`@cql_input`)?

File: graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/MutationMappingModelBuilder.java (hunk 4 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/MutationMappingModelBuilder.java:43]: No, this is `graphql.language.Type`. It represents a type reference in the AST, with only 3 possible subtypes:
* `TypeName`: a type itself, like `User`, `Boolean`
* `NonNullType`: `User!`
* `ListType`: `[User]`

File: graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/MutationMappingModelFactory.java (hunk 5 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/MutationMappingModelFactory.java:76]: Selects are queries, not mutations. We process queries separately, see `MappingModel.build()`.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/ResponseMappingModel.java (hunk 6 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/ResponseMappingModel.java:86]: `PAGING_STATE` definitely (not relevant for inserts, but response payloads will be possible with selects too).

I thought there would be others, but looking at what is available on `ResultSet`, that might be it.

File: testing/src/test/java/io/stargate/it/http/graphql/graphfirst/InsertTest.java (hunk 7 above)
Comment (olim7t) [reply testing/src/test/java/io/stargate/it/http/graphql/graphfirst/InsertTest.java:53]: We can't predict the contents though: depending on its intended use, a response payload may wrap a single entity or a list, and may or may not have technical fields. It's even possible to have just the `applied` field if the caller just needs to know if a LWT failed or not, but is not interested in the existing data.

I guess we could encode that information with directive parameters:
```
type User
@cql_payload(name: "InsertUserResponse", entityField: SINGLE, technicalFields: APPLIED)
{
  ...
}
```
But overall I'm not sure if this is an improvement: it's almost as long to type, and it's less clear what's going on.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/MutationMappingModelBuilder.java (hunk 4 above)
Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/MutationMappingModelBuilder.java:43]: How maps are represented?

File: testing/src/test/java/io/stargate/it/http/graphql/graphfirst/InsertTest.java (hunk 7 above)
Comment (tomekl007) [reply testing/src/test/java/io/stargate/it/http/graphql/graphfirst/InsertTest.java:53]: I agree that the current solution is better

File: graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/MutationMappingModelBuilder.java (hunk 4 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/MutationMappingModelBuilder.java:43]: There are no maps in GraphQL: https://spec.graphql.org/June2018/#sec-Wrapping-Types.

**Pull request merged**